File: src/application/useCases/loadCashSummary.ts
import { summarizeCash, type CashSummary } from '../../domain/cashSummary'
import type { Period } from '../../domain/period'
import type { AllocationRepository, TransactionRepository, WalletRepository } from '../ports/repositories'

/**
 * LoadCashSummary — كارت الكاش في الرئيسية وتفاصيله (OVERRIDES §32، رد المالك «الاتنين»).
 * `null` = الحساب مالوش محفظة كاش (مش «صفر»، قاعدة 10).
 *
 * ⚠️ تكلفة القراءة: استعلام واحد من أقدم تاريخ (بداية الكاش أو بداية الفترة) لحد آخر الفترة أو النهارده.
 * محفظة كاش بدايتها قديمة جدًا معناها قراءة أكتر — مقبول لأن رصيد البداية بيتحدث من أسئلة البداية والإعدادات.
 */
export function makeLoadCashSummary(deps: {
  wallets: WalletRepository
  txns: TransactionRepository
  allocations: AllocationRepository
}) {
  return async function load({ period, today }: { period: Period; today: string }): Promise<CashSummary | null> {
    const wallet = (await deps.wallets.listAll()).find((w) => w.kind === 'cash') ?? null
    if (!wallet) return null
    const from = wallet.openingAt < period.start ? wallet.openingAt : period.start
    const to = today > period.end ? today : period.end
    const transactions = await deps.txns.listByDateRange(from, to)
    const inPeriod = transactions.filter((t) => t.occurredAt >= period.start && t.occurredAt <= period.end)
    const allocations = await deps.allocations.listByTransactionIds(inPeriod.map((t) => t.id))
    return summarizeCash({ wallet, transactions, allocations, period })
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/usecase/LoadCashSummary.kt
package app.masroufy.usecase

import app.masroufy.core.CashSummary
import app.masroufy.core.Period
import app.masroufy.core.summarizeCash
import app.masroufy.core.withEstimatedKinds
import app.masroufy.port.AllocationRepository
import app.masroufy.port.CategoryRepository
import app.masroufy.port.TransactionRepository
import app.masroufy.port.WalletRepository

/**
 * LoadCashSummary — نقل `loadCashSummary.ts`: كارت الكاش في الرئيسية وتفاصيله (OVERRIDES §32).
 * `null` = الحساب مالوش محفظة كاش (مش «صفر»، قاعدة 10). العمليات بتتحسب بنفس الأنواع
 * التقديرية اللي بيستعملها رقم «ما تم صرفه» (OVERRIDES §18).
 */

data class LoadCashSummaryDeps(
    val wallets: WalletRepository,
    val txns: TransactionRepository,
    val allocations: AllocationRepository,
    val categories: CategoryRepository,
)

class LoadCashSummary(private val deps: LoadCashSummaryDeps) {
    suspend fun load(period: Period, today: String): CashSummary? {
        val wallet = deps.wallets.listAll().find { it.kind == "cash" } ?: return null
        val from = if (wallet.openingAt < period.start) wallet.openingAt else period.start
        val to = if (today > period.end) today else period.end
        val rows = deps.txns.listByDateRange(from, to)
        val categories = deps.categories.listAll()
        val transactions = withEstimatedKinds(rows, categories.associate { it.id to it.name }).transactions
        val inPeriod = transactions.filter { it.occurredAt >= period.start && it.occurredAt <= period.end }
        val allocations = deps.allocations.listByTransactionIds(inPeriod.map { it.id })
        return summarizeCash(wallet, transactions, allocations, period.start, period.end)
    }
}
